crates/rust-analyzer/src/handlers.rs

```diff
@@ -1916,3 +1939,24 @@ fn run_rustfmt(
         Ok(Some(to_proto::text_edit_vec(&line_index, diff(&file, &new_text))))
     }
 }
+
+//Thats a best effort to try and find the crate path
+fn crate_path(crate_name: Option<&String>, root_file_path: VfsPath) -> Option<VfsPath> {
+    crate_name.and_then(|crate_name| {
+        let mut crate_path = None;
+        let mut root_path = root_file_path;
+        while let Some(path) = root_path.parent() {
+            match path.name_and_extension() {
+                Some((name, _)) => {
+                    if name.starts_with(crate_name.as_str()) {
+                        crate_path = Some(path);
+                        break;
+                    }
+                }
+                None => break,
+            }
+            root_path = path;
+        }
+        crate_path
+    })
+}
```

Comment:
This at least deserves a comment explaining what it does as it is not clear at all. Though we could be smarter about this I think. 
We can iterate over all cargo workspaces we know, iterate over the pacakages there, then iterate over the targets that belong to those packages and check if the targets root file path is the same as the path we are looking for. We now have a package with a fitting root path, from there we can read out the cargo toml path of the package and go up to the dir.

(this is obviously quadratic so we probably want to do some upfront hashmap collecting once to reduce the iteration amount required.